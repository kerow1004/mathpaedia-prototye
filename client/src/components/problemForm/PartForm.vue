/* global MathJax */

<template>
  <div>
    <!-- Modal Component -->
    <b-modal ref="mathEditor"
             title="Math Editor"
             @ok="handleOk">
      <quill v-model="content" :config="config" @text-change="textChange"></quill>
    </b-modal>

    <b-form-group id="problemSentenceGroup"
                  label="Sentence:"
                  description="Click to edit sentence with math equations">
      <div id="problemSentence" class="input-data"
           @click="openMathEditor">{{ part.sentence }}
      </div>
    </b-form-group>

    <b-form-group id="problemPictureGroup"
                  label="Picture(Graph):"
                  label-for="problemPicture">
      <b-form-input id="problemPicture"
                    type="text"
                    v-model="part.picture"
                    placeholder="Enter URL of math picture(graph)">
      </b-form-input>
    </b-form-group>

    <b-form-group id="problemChoicesGroup"
                  label="Choices:">
      <div v-for="(choice, index) in part.choices" :key="index" class="input-choice">
        <b-input-group :prepend="`Choice ${index+1}`">
          <b-form-input type="text"
                        v-model="choice.value">
          </b-form-input>
          <b-input-group-append>
            <b-btn variant="danger"
                   @click="removeChoice(index)"
                   v-if="part.choices.length > 1">-</b-btn>
            <b-btn variant="success"
                   @click="addMoreChoice">+</b-btn>
          </b-input-group-append>
        </b-input-group>
      </div>
    </b-form-group>
  </div>
</template>

<script>
import Vue from 'vue';
import VueQuill from 'vue-quill';

Vue.use(VueQuill);

export default {
  name: 'PartForm',
  props: ['part'],
  data() {
    return {
      content: {
        ops: [],
      },
      config: {
        modules: {
          formula: true,
          toolbar: [
            ['bold', 'italic', 'underline'],
            ['image', 'code-block'],
            ['formula'],
          ],
        },
      },
      show: true,
    };
  },
  methods: {
    addMoreChoice() {
      this.part.choices.push({ value: '' });
    },
    removeChoice(index) {
      this.part.choices.splice(index, 1);
    },
    openMathEditor() {
      this.$log.info('openMathEditor');
      this.$refs.mathEditor.show();
    },
    textChange(editor) {
      // this.$log.info('[editor/text-change] Text', editor.getText());
      this.$log.info('[editor/text-change] Content', editor.getContents());
    },
    handleOk() {
      const { ops } = this.content;

      this.part.sentence = '';

      ops.forEach((op) => {
        if (this._.isPlainObject(op.insert)) {
          if (op.insert.formula) {
            this.part.sentence += `$${op.insert.formula}$`;
          }
        } else {
          this.part.sentence += op.insert;
        }
      });

      this.$log.info('combined sentence', this.part.sentence);

      /* eslint-disable */
      this.$nextTick(() => {
        MathJax.Hub.Queue(['Typeset', MathJax.Hub, 'content-sentence']);
      });
      /* eslint-enable */
    },
  },
};
</script>

<style scoped>
  .input-data {
    border: 1px solid black;
    padding: 10px;
    white-space: pre-wrap;
  }

  .input-choice {
    margin: 5px;
  }
</style>
