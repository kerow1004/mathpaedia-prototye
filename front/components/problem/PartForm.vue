/* global MathJax */

<template>
  <div>
    <!-- Modal Component -->
    <v-dialog v-model="showEditor">
      <v-card>
        <v-card-title>Math Editor</v-card-title>
        <v-card-text>
          <!--<quill v-model="content" :config="config" @text-change="textChange"></quill>-->
          <div class="quill-editor"
               :content="content"
               v-quill:myQuillEditor="config">
          </div>
        </v-card-text>
        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn color="green darken-1" flat @click.native="handleCancel">Cancel</v-btn>
          <v-btn color="green darken-1" flat @click.native="handleOk">OK</v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>

    <v-container>
      <v-layout row>
        <v-flex xs4>
          <v-subheader>Sentence</v-subheader>
        </v-flex>
        <v-flex xs8>
          <div id="problemSentence" class="input-data"
               @click="openMathEditor">{{ part.sentence }}
          </div>
        </v-flex>
      </v-layout>

      <v-layout row>
        <v-flex xs4>
          <v-subheader>Picture(Graph)</v-subheader>
        </v-flex>
        <v-flex xs8>
          <v-text-field
            v-model="part.picture"
            placeholder="Enter URL of math picture(graph)"
          ></v-text-field>
        </v-flex>
      </v-layout>

      <v-layout>
        <v-flex xs4>
          <v-subheader>Choices</v-subheader>
        </v-flex>
        <v-flex xs8>
          <div v-for="(choice, index) in part.choices" :key="index" class="input-choice">
            <v-layout>
              <v-flex xs10>
                Choice {{ index+1 }}
                <v-text-field
                  v-model="choice.value"
                ></v-text-field>
              </v-flex>
              <v-flex xs2>
                <v-btn small color="warning"
                       @click="removeChoice(index)"
                       v-if="part.choices.length > 1">-</v-btn>
                <v-btn small color="primary"
                       @click="addMoreChoice">+</v-btn>
              </v-flex>
            </v-layout>
          </div>
        </v-flex>
      </v-layout>
    </v-container>
  </div>
</template>

<script>
export default {
  name: 'PartForm',
  props: ['part'],
  data() {
    return {
      content: '',
      config: {
        modules: {
          formula: true,
          toolbar: [
            ['bold', 'italic', 'underline'],
            ['image', 'code-block'],
            ['formula'],
          ],
        },
      },
      showEditor: false,
    };
  },
  methods: {
    addMoreChoice() {
      this.part.choices.push({ value: '' });
    },
    removeChoice(index) {
      this.part.choices.splice(index, 1);
    },
    openMathEditor() {
      this.showEditor = true;
    },
    handleOk() {
      const ops = this.myQuillEditor.getContents();

      this.showEditor = false;

      this.part.sentence = '';

      ops.forEach((op) => {
        if (_.isPlainObject(op.insert)) {
          if (op.insert.formula) {
            this.part.sentence += `$${op.insert.formula}$`;
          }
        } else {
          this.part.sentence += op.insert;
        }
      });
    },
    handleCancel() {
      this.showEditor = false;
    },
  },
};
</script>

<style scoped>
  .input-data {
    border: 1px solid black;
    padding: 10px;
    white-space: pre-wrap;
  }

  .input-choice {
    margin: 5px;
  }
</style>
