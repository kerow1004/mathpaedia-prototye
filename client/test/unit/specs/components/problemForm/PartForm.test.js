import { mount, createLocalVue } from '@vue/test-utils';
import VueLogger from 'vuejs-logger';
import BootstrapVue from 'bootstrap-vue';
import PartForm from '@/components/problemForm/PartForm';

const localVue = createLocalVue();
localVue.use(VueLogger);
localVue.use(BootstrapVue);

describe('PartForm.vue', () => {
  const part = {
    sentence: '',
    picture: '',
    choices: [
      {
        value: '',
      },
    ],
  };

  const oriMathJax = global.MathJax;

  let wrapper;

  beforeAll(() => {
    global.MathJax = {
      Hub: {
        Queue: jest.fn(),
      },
    };
  });

  afterAll(() => {
    global.MathJax = oriMathJax;
  });

  beforeEach(() => {
    wrapper = mount(PartForm, {
      propsData: {
        part,
      },
      localVue,
      stubs: ['quill'],
    });
  });

  test('has the expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('remove one choice when removeChoice is clicked', () => {
    const addButton = wrapper.find('.input-choice button');
    const buttonCount = wrapper.findAll('button').length;
    addButton.trigger('click');

    const removeButton = wrapper.find('.input-choice button');
    removeButton.trigger('click');

    const buttonCountAfterAdd = wrapper.findAll('button').length;
    expect(buttonCountAfterAdd).toBe(buttonCount);
  });
});
