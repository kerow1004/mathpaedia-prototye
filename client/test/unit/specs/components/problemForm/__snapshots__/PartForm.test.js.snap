// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartForm.vue has the expected html structure 1`] = `
<div>
  <div>
    <div
      id="__BVID__2___BV_modal_outer_"
    >
      <transition-stub />
      <!---->
    </div>
  </div>
   
  <fieldset
    aria-describedby="problemSentenceGroup__BV_description_"
    aria-labelledby="problemSentenceGroup__BV_label_"
    class="b-form-group form-group"
    id="problemSentenceGroup"
    role="group"
  >
    <legend
      class="col-form-label pt-0"
      id="problemSentenceGroup__BV_label_"
    >
      Sentence:
    </legend>
    <div
      aria-labelledby="problemSentenceGroup__BV_label_"
      class=""
      role="group"
    >
      <div
        class="input-data"
        id="problemSentence"
      >
        
    
      </div>
      <!---->
      <!---->
      <small
        class="form-text text-muted"
        id="problemSentenceGroup__BV_description_"
      >
        Click to edit sentence with math equations
      </small>
    </div>
  </fieldset>
   
  <div
    aria-labelledby="problemPictureGroup__BV_label_"
    class="b-form-group form-group"
    id="problemPictureGroup"
    role="group"
  >
    <label
      class="col-form-label pt-0"
      for="problemPicture"
      id="problemPictureGroup__BV_label_"
    >
      Picture(Graph):
    </label>
    <div
      class=""
    >
      <input
        class="form-control"
        id="problemPicture"
        placeholder="Enter URL of math picture(graph)"
        type="text"
        value=""
      />
      <!---->
      <!---->
      <!---->
    </div>
  </div>
   
  <fieldset
    aria-labelledby="problemChoicesGroup__BV_label_"
    class="b-form-group form-group"
    id="problemChoicesGroup"
    role="group"
  >
    <legend
      class="col-form-label pt-0"
      id="problemChoicesGroup__BV_label_"
    >
      Choices:
    </legend>
    <div
      aria-labelledby="problemChoicesGroup__BV_label_"
      class=""
      role="group"
    >
      <div
        class="input-choice"
      >
        <div
          class="input-group"
          role="group"
        >
          <div
            class="input-group-prepend"
          >
            <div
              class="input-group-text"
            >
              Choice 1
            </div>
          </div>
          <input
            class="form-control"
            id="__BVID__8"
            type="text"
            value=""
          />
           
          <div
            class="input-group-append"
          >
            <!---->
             
            <button
              class="btn btn-success"
              type="button"
            >
              +
            </button>
          </div>
        </div>
      </div>
      <!---->
      <!---->
      <!---->
    </div>
  </fieldset>
</div>
`;
